cumulus_library/studies/core/core_templates/core_utils.jinja

```diff
@@ -1,38 +1,138 @@
+{#- A collection of macros for constructing common safe column selects
+
+    For layout, is is generally assumed that these are being invoked from within
+    a WITH TABLE AS(...) statement, so they will always assume two levels of
+    indentation.
+-#}
+
 {%- macro comma_delineate(loop) -%}
 {%- if not loop.last -%}
 , 
 {%- endif -%}
 {%- endmacro -%}
 
--- gets a list of columns from a known present source
+{#- gets a list of columns from a known present source -#}
 {% macro basic_cols(table, alias, targets) %}
         {%- for col in targets %}
         {{ alias }}.{{ col.lower() }}
         {{- comma_delineate(loop) }}
         {%- endfor %}
 {%- endmacro %}
 
+{#- gets a list of columns from a potentially null source.
+
+targets is an array expecting a data type of the following:
+    string: refers to a vanilla SQL primitive column
+    tuple(col, object, col_alias): refers to a one level nested SQL object
+    tuple(col, object, , child, col_alias): refers to a two level nested SQL object
 
--- gets a column, if present, from a basic SQL primitive col, or a null otherwise
-{% macro basic_nullable_cols(table, alias, targets, schema) %}
+-#}
+{% macro nullable_cols(table, alias, targets, schema) %}
         {%- for col in targets %}
-        {%- if schema[table][col] %}
+        {#- depth one nested column-#}
+        {%- if col is not string and col|length ==3%}
+        {%- if schema[table][col[0]][col[1]] %}
+        {{ alias }}.{{ col[0].lower() }}.{{ col[1].lower() }} AS {{col[2]}}
+        {%- else %}
+        NULL AS {{col[2]}}
+        {%- endif %}
+        {#- depth two nested column -#}
+        {%- elif col is not string and col|length ==4%}
+        {%- if schema[table][col[0]][col[2]] %}
+        {{ alias }}.{{ col[0].lower() }}.{{ col[1].lower() }}.{{ col[2].lower() }} AS {{col[3]}}
```

Comment:
So I was inspired by this line to look up whether SQL does provide guarantees about case insensivity for column names. Looks like (as always with SQL) [everything depends](https://learnsql.com/blog/sql-case-sensitive/). Mostly things are case insensitive, but the default collation on some SQL servers will use the system default (so be case sensitive on Linux and insensitive on Windows - yikes). But collation can be changed.

Anyway, I'm skeptical on our need to cater to oddball databases that we don't even support right now, but I'm fine with it. Our two database targets both are insensitive for column names. It might be interesting to find out if Azure is (I would have to assume it is, but worth validating).

Reply:
so - this is actually not about catering to a specific DB. the intent here is to make information_schema.cols-style calls easier for downstream users, since we hide it behind some templates they can elect to use. the goal is to be able to use the results of that table as a means to look into other sources without having to do casts. 

if that's sensible, i can document the reason for this rats nest for future readers? i agree its ugly as hell, but i like the end product.

Reply:
I didn't follow your intent comment - maybe let's chat on slack or something for higher bandwidth clarity

Reply:
From that discussion: Matt was talking about new (alias) column names, I meant to be talking about the column name _lookups_.

We both agreed a single standard made sense for aliases.